examples/cis3-nft-sponsored-txs/tests/tests.rs

```diff
@@ -19,11 +26,419 @@ const ACC_INITIAL_BALANCE: Amount = Amount::from_ccd(10000);
+        },
+        signer:    ALICE,
+        message:   PermitMessage {
+            timestamp:        Timestamp::from_timestamp_millis(10000000000),
+            contract_address: ContractAddress {
+                index:    0,
+                subindex: 0,
+            },
+            entry_point:      OwnedEntrypointName::new_unchecked("transfer".into()),
+            nonce:            0,
+            payload:          to_bytes(&payload),
+        },
+    };
+
+    // Get the message hash to be signed.
+    let invoke = chain
+        .contract_invoke(BOB, BOB_ADDR, Energy::from(10000), UpdateContractPayload {
+            amount:       Amount::zero(),
+            address:      contract_address,
+            receive_name: OwnedReceiveName::new_unchecked("cis3_nft.viewMessageHash".to_string()),
+            message:      OwnedParameter::from_serial(&permit_transfer_param)
+                .expect("Should be a valid inut parameter"),
+        })
+        .expect("Should be able to query viewMessageHash");
+
+    let message_hash: HashSha2256 =
+        from_bytes(&invoke.return_value).expect("Should return a valid result");
+
+    permit_transfer_param.signature = keypairs
+        .expect("Should have a generated private key to sign")
+        .sign_message(&to_bytes(&message_hash));
+
+    // Transfer token with the permit function.
+    let update = chain
+        .contract_update(
+            Signer::with_one_key(),
+            BOB,
+            BOB_ADDR,
+            Energy::from(10000),
+            UpdateContractPayload {
+                amount:       Amount::zero(),
+                address:      contract_address,
+                receive_name: OwnedReceiveName::new_unchecked("cis3_nft.permit".to_string()),
+                message:      OwnedParameter::from_serial(&permit_transfer_param)
+                    .expect("Should be a valid inut parameter"),
+            },
+        )
+        .expect("Should be able to transfer token with permit");
+
+    // Check that the correct events occurred.
+    let events = update
+        .events()
+        .flat_map(|(_addr, events)| events.iter().map(|e| e.parse().expect("Deserialize event")))
+        .collect::<Vec<Event>>();
+
+    assert_eq!(events, [
+        Event::Cis2Event(Cis2Event::Transfer(TransferEvent {
+            token_id: TOKEN_1,
+            amount:   ContractTokenAmount::from(1),
+            from:     ALICE_ADDR,
+            to:       BOB_ADDR,
+        })),
+        Event::Nonce(NonceEvent {
+            account: ALICE,
+            nonce:   0,
+        })
+    ]);
+
+    // Check balances in state.
+    let balance_of_alice_and_bob = get_balances(&chain, contract_address);
+
+    assert_eq!(balance_of_alice_and_bob.0, [TokenAmountU8(0), TokenAmountU8(1)]);
+}
+
+/// Test permit transfer function. The signature is generated outside this test
+/// case (e.g. with https://cyphr.me/ed25519_tool/ed.html). TOKEN_1 is transferred from Alice to Bob.
```

Comment:
I don't understand why the test is written this way. It adds complications. Why don't we write it in the way that we recommend, generating signatures inside the tests themselves.

Reply:
I agree that the recommended way would be to pick either to generate all signatures in the test case (which should be the default way) or generate all signatures outside of the test cases.

This is an example contract providing both test case options, because sometimes you still need hardcoded signatures (e.g. if you have a setup such as the umbrella smart contracts where production account addresses are hardcoded in the smart contract, in that case, you don't want or don't can expose your private key in the smart contract testing file).